chromadb/rate_limiting/admission_controller.py

```diff
@@ -0,0 +1,358 @@
+import threading
+import time
+import concurrent.futures
+import queue
+
+from typing import Optional, cast
+
+
+class Ticket:
+    def __init__(self, issued_by: Optional["AdmissionController"]) -> None:
+        self.acquisition_elapsed = 0.0
+        self.issued_by = issued_by
+
+    def release(self) -> None:
+        if self.issued_by:
+            self.issued_by.release()
+            self.issued_by = None
+
+
+class Waiter:
+    def __init__(self) -> None:
+        self.index = 0
+        self.woke_up = False
+
+    def notify(self) -> None:
+        self.woke_up = True
+
+
+class AdmissionController:
```

Comment:
Is this a common pattern in our codebase?  I'd prefer just one class that implements the AC unless we have other AC impls.  In which case, we should probably use the abstract base class support built into Python.

Reply:
I think you will want this to be a `Component` - https://github.com/chroma-core/chroma/blob/e2628e819165442bc68dea1f168c6b8d676c8744/chromadb/config.py#L284 To match the pattern of the rest of the codebase.

This lets us dependency inject the rate limiting, so we can have a no-op impl.